runtime/near-vm-logic/src/logic.rs

```diff
@@ -943,14 +945,104 @@ impl<'a> VMLogic<'a> {
     pub fn keccak512(&mut self, value_len: u64, value_ptr: u64, register_id: u64) -> Result<()> {
         self.gas_counter.pay_base(keccak512_base)?;
         let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
-        self.gas_counter.pay_per_byte(keccak512_byte, value.len() as u64)?;
+        self.gas_counter.pay_per(keccak512_byte, value.len() as u64)?;
 
         use sha3::Digest;
 
         let value_hash = sha3::Keccak512::digest(&value);
         self.internal_write_register(register_id, value_hash.as_slice().to_vec())
     }
 
+    /// Hashes the given value using RIPEMD-160 and returns it into `register_id`.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    ///  Where `message_blocks` is `(value_len + 9).div_ceil(64)`.
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + ripemd160_base + ripemd160_block * message_blocks`
+    #[cfg(feature = "protocol_feature_math_extension")]
+    pub fn ripemd160(&mut self, value_len: u64, value_ptr: u64, register_id: u64) -> Result<()> {
+        self.gas_counter.pay_base(ripemd160_base)?;
+        let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
+
+        let message_blocks = value
```

Comment:
> @birchmd You ceiling division code is still too complex. Something like `let message_blocks = (value.len() as u64).checked_add(8).ok_or(VMLogicError::HostError(HostError::IntegerOverflow))? + 1;` should work (and not need the `div_ceil` function).

@abacabadabacaba I don't understand your simplification, where did the division by 64 go? If the division by 64 is incorrect logic (I don't know if it is or not I was only looking to address superficial comments) then of course we don't need any division function, but that's not what the original comment was about.

Reply:
Sorry, fixed the snippet.

Reply:
Is it ok to always have the `+ 1` even in the case that `value.len() + 9` is evenly divisible by 64? Technically that is not ceiling division because if the result is evenly divisible then we shouldn't need the `+ 1`, but if you think that's ok then I'm happy to accept the simplification.

Reply:
That's why I add 8 instead of 9.
(len+9).ceiling_div(64) = (len+9+63) / 64 = (len+8+64) / 64 = (len+8) / 64 + 1

Reply:
Ah, makes sense. Thanks!

Done in https://github.com/near/nearcore/pull/4380/commits/aa1b8c5321594be675a28eedab1679759523689b

Reply:
(and `div_ceil` function removed in https://github.com/near/nearcore/pull/4380/commits/1ec6416efb9f6b62299aa170524561b4d35cbf4d )